src/django_rich/test.py

```diff
@@ -0,0 +1,157 @@
+
+    def addSuccess(self, test: TestCase) -> None:
+        if self.showAll:
+            self.console.print("ok", style=DJANGO_GREEN)
+        elif self.dots:
+            self.console.print(".", style=DJANGO_GREEN, end="")
+
+    @failfast
+    def addError(self, test: TestCase, err: _SysExcInfoType) -> None:
+        self.errors.append((test, self._exc_info_to_string(err, test)))
+        self._mirrorOutput = True
+        if self.showAll:
+            self.console.print("ERROR", style=Style(color="red"))
+        elif self.dots:
+            self.console.print("E", style=Style(color="red"), end="")
+
+    @failfast
+    def addFailure(self, test: TestCase, err: _SysExcInfoType) -> None:
+        self.failures.append((test, self._exc_info_to_string(err, test)))
+        self._mirrorOutput = True
+        if self.showAll:
+            self.console.print("FAIL", style=Style(color="red"))
+        elif self.dots:
+            self.console.print("F", style=Style(color="red"), end="")
+
+    def addSkip(self, test: TestCase, reason: str) -> None:
+        self.skipped.append((test, reason))
+        if self.showAll:
+            self.console.print(f"skipped {reason!r}", style=Style(color="yellow"))
+        elif self.dots:
+            self.console.print("s", style=Style(color="yellow"), end="")
+
+    def addExpectedFailure(self, test: TestCase, err: _SysExcInfoType) -> None:
+        self.expectedFailures.append((test, self._exc_info_to_string(err, test)))
+        if self.showAll:
+            self.console.print("expected failure", style=Style(color="yellow"))
+        elif self.dots:
+            self.console.print("x", style=Style(color="yellow"), end="")
+
+    @failfast
+    def addUnexpectedSuccess(self, test: TestCase) -> None:
+        self.unexpectedSuccesses.append(test)
+        if self.showAll:
+            self.console.print("unexpected success", style=Style(color="red"))
+        elif self.dots:
+            self.console.print("u", style=Style(color="red"), end="")
+
+    # Typeshed had wrong signature
+    # https://github.com/python/typeshed/pull/7341
+    def printErrorList(  # type: ignore [override]
+        self,
+        flavour: str,
+        errors: list[tuple[TestCase, str]],
+    ) -> None:
+        for test, err in errors:
+            rule = Rule(style=DJANGO_GREEN)
+            title = f"{flavour}: {self.getDescription(test)}"
+            self.console.print(rule, title, rule)
+            self.stream.write("%s\n" % err)
+
+    def _exc_info_to_string(self, err: _SysExcInfoType, test: TestCase) -> str:
+        """Converts a sys.exc_info()-style tuple of values into a string."""
+        exctype, value, tb = err
+        # Skip test runner traceback levels
+        while tb and self._is_relevant_tb_level(tb):  # type: ignore [attr-defined]
+            tb = tb.tb_next
+
+        msgLines = []
+        if exctype is not None:
+            assert value is not None
+            assert tb is not None
```

Comment:
I added these asserts for type narrowing purposes, because Rich's type annotations require an exception to be used. I think the `tb is not None` is too strict though as a non-existent test triggers a crash:

```
$ DJANGO_SETTINGS_MODULE=tests.settings .tox/py310-django40/bin/python -m django test does_not_exist
System check identified no issues (0 silenced).
Traceback (most recent call last):
  ...
  File "/.../3.10.2/lib/python3.10/unittest/runner.py", line 184, in run
    test(result)
  File "/.../3.10.2/lib/python3.10/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/.../3.10.2/lib/python3.10/unittest/suite.py", line 122, in run
    test(result)
  File "/.../3.10.2/lib/python3.10/unittest/case.py", line 650, in __call__
    return self.run(*args, **kwds)
  File "/.../3.10.2/lib/python3.10/unittest/case.py", line 599, in run
    self._feedErrorsToResult(result, outcome.errors)
  File "/.../3.10.2/lib/python3.10/unittest/case.py", line 518, in _feedErrorsToResult
    result.addError(test, exc_info)
  File "/.../3.10.2/lib/python3.10/unittest/result.py", line 17, in inner
    return method(self, *args, **kw)
  File "/.../python3.10/site-packages/django_rich/test.py", line 50, in addError
    self.errors.append((test, self._exc_info_to_string(err, test)))
  File "/.../python3.10/site-packages/django_rich/test.py", line 111, in _exc_info_to_string
    assert tb is not None
AssertionError
```

Reply:
So turns out that yes, `tb` will be `None` for such a case. I've removed that assertion and added a test for bad test path.